Refactor windowing processors into single class

viliam-durina opened the pull request:
Refactor windowing processors into single class

GroupByFrameP, SlidingWindowP and recently also TumblingWindowP were very 
similar. They had the same internal state, they processed either events or 
Frames on input, and their output Frames contained either accumulator or 
finished result. Here, they are refactored into single class.

Advantages of this:
- less code
- uses HashMap instead of TreeMap (a bit faster)
- as a bonus, it allowed us to do one stage sliding window aggregation, which 
might be needed, if there are too many keys to fit into single machine.

devOpsHazelcast commented:
Test FAILed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test FAILed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test FAILed.

devOpsHazelcast commented:
Test PASSed.

mtopolnik commented on hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/WindowingProcessors.java:
```diff
@@ -17,13 +17,58 @@
+ * only pre-accumulated results are sent over the network. This greatly
+ * reduces the amount of data sent over the network. The DAG is:
+ *
+ * <pre>{@code
```
Since you don't use `<` in the diagram, you don't need `{@code}`

mtopolnik commented on hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/WindowingProcessors.java:
```diff
@@ -17,13 +17,58 @@
 /**
  * Contains factory methods for processors dealing with windowing operations.
+ *
+ * <h1>Two-step aggregation</h1>
```
We should use the term "stage" consistently.

mtopolnik commented on hazelcast-jet-core/src/test/java/com/hazelcast/jet/windowing/SlidingWindowIntegrationTest.java:
```diff
@@ -44,24 +49,33 @@
 public class SlidingWindowIntegrationTest extends JetTestSupport {
 
+    @Parameter
+    public boolean oneStageProcessor;
```
Should be renamed to "singleStageProcessor"

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

cangencer commented on hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/WindowingProcessors.java:
```diff
@@ -47,59 +93,155 @@ private WindowingProcessors() {
+            @Nonnull WindowOperation<? super T, A, ?> windowOperation
     ) {
-        return groupByFrame(t -> "global", extractEventSeqF, windowDef, collector);
+        return slidingWindowStage1(t -> "global", extractEventSeqF, windowDef, windowOperation);
```
I would prefer `ALL` instead of "global". Also it should be a static variable.

mtopolnik commented on hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/WindowingProcessors.java:
```diff
@@ -47,59 +93,155 @@ private WindowingProcessors() {
+                tumblingWinDef,
+                item -> tumblingWinDef.higherFrameSeq(extractEventSeqF.applyAsLong(item)),
+                extractKeyF,
+                WindowOperation.of(
```
Another thing we can have is a windowOperation that takes a `BiConsumer accumulateItemF` and wraps it into a `BiFunction`. There are already many places where this is done by hand.

viliam-durina replied:
I wouldn't bother with an overload, they are pain to maintain and writing `a -> a` is trivial. Normally, mutable accumulator is the way to go, due to much less GC load. I would even consider using `BiConsumer` for all `accumulateItem`, `combineAccumulators` and `deductAccumulators`, this will remove a lot of code. And if one wants an immutable container, he will use `WindowingOperations.reducing`, which even doesn't have a finisher. In next PR, if you wish.

mtopolnik replied:
It's not `a -> a`; it's `(acc, item) -> { biConsumer.accept(acc, item); return biConsumer; }` instead of simply `biConsumer`. You may be right on the point that the separation of the two cases (mutable/immutable) could go even deeper.

viliam-durina commented:
I addressed your comments. Additionally, I renamed tests. I removed the `TumblingWindowPTest`, which was essentially a copy of `GroupByFramePTest`, but with finished accumulator value.

devOpsHazelcast commented:
Test PASSed.
